fixed:issue->#1835 Support NYNAB import of transactions that contain subtransactions tha…

Marethyu1 opened the pull request:
…t are transfers

A subtransaction is considered a transfer from ynab if
- the transfer_account_id is set
- the category_id is null (transfers dont have categories)

This addresses #1835 

<!-- Thank you for submitting a pull request! Make sure to follow the instructions to write release notes for your PR — it should only take a minute or two: https://github.com/actualbudget/docs#writing-good-release-notes -->

MatissJanis commented:
Good catch. I like this fix, but I'm a bit worried about merging it as it exposes a different issue: sub-transaction transfers do not appear correctly.

Reproduction:
1. use the demo budget in the issue ([ynab5-demo-budget.json](https://github.com/actualbudget/actual/files/13257078/ynab5-demo-budget.json))
2. import it
3. observe "200" in the "Saving" account
4. expectation: the account has 250

So basically 50$ just disappeared from the budget. That's definitely not something we want.

I wonder if we should fix this by disallowing transfers in sub-transactions. 🤔

Marethyu1 commented:
I suspect that this is a mistake on my part - I'll take a look. The reason I believe that is
- I would also expect the savings account to have 250
- I updated the onboarding test to expect 250 in the savings account and it passes (this is what suprises me)

Marethyu1 commented:
Ah so, the initial test budget I uploaded is wrong. I have updated the issue with a correct budget, and also you can check It out here [ynab5-demo-budget.json](https://github.com/actualbudget/actual/files/13257675/ynab5-demo-budget.json)

The problem with the original budget I uploaded was that it was missing the corresponding transaction in savings account of +50 coming into the savings account. 

The reason why the tests were passing on this pr was that I fixed the issue in the pull request but didn't go back and update the linked budget in #1835 

TLDR: The changes in this pr should now match the new demo budget in #1835. These should now result in the savings account to have 250

MatissJanis commented on packages/loot-core/src/server/importers/ynab5.ts:
```diff
@@ -178,11 +178,21 @@ async function importTransactions(
             subtransactions: subtransactions
               ? subtransactions.map(subtrans => {
+                  let payee = null;
+                  if (subtrans.transfer_account_id) {
+                    payee = payees.find(
+                      p =>
+                        p.transfer_acct ===
+                        entityIdMap.get(subtrans.transfer_account_id),
+                    ).id;
+                  }
```
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**💬 suggestion:** ‏I suspect for bigger imports this operation will have poor performance. Would you mind pre-computing a hashmap for payees and then just using it here as a lookup?

i.e.
```
const payeeTransferActHashMap = new Map(payees.map((payee) => [payee.transfer_acct, payee]));

// ... and then inside the loop
payee = payeeTransferActHashMap[entityIdMap.get(subtrans.transfer_account_id)]?.id;

Marethyu1 replied:
Updated based on feedback :) let me know if this isnt quite what you were meaning :)

MatissJanis commented on packages/loot-core/src/server/importers/ynab5.ts:
```diff
@@ -178,11 +178,21 @@ async function importTransactions(
                     amount: amountFromYnab(subtrans.amount),
                     category: entityIdMap.get(subtrans.category_id) || null,
                     notes: subtrans.memo,
+                    payee: payee,
```
```suggestion
                    payee,
```

Marethyu1 replied:
Added 👍

MatissJanis reviewed (changes requested):
Perfect! I was able to test it and it works as expected now. Thanks for clarifying!

Would you mind doing this one small perf improvement? Performance for imports is quite important as some folks come to Actual with years of data (so 100k+ transactions).

MatissJanis reviewed (approved):
Perfect! Thanks so much for adding this improvement.
